Overhaul for upstream-reproducible tarballs

iskunk opened the pull request:
This change makes it possible to generate primary and testdata tarballs that match Google's lite and testdata tarballs bit-perfectly. First confirmed with the 132.0.6834.110 release.

> **Non-commit-message note:** This is the big one, been working on this for a while. Please try it out, confirm that you get bit-identical tarballs, and let me know if you see any issues. Once this is in, we may want to consider having a scheduled workflow that just verifies the match between our tarball and Google's.

New file `check.patch`: This patch is not meant to modify any files; it is used only to verify that a critical section of the upstream `publish_tarball.py` script hasn't changed on us. If it causes a `patch --dry-run` invocation to fail, then that indicates an update to `package_chromium.sh` is probably needed.

Delete `export_tarball.py`: We now pull down the upstream Git repo that contains this script, and run it there.

Numerous changes in `package_chromium.sh`:

* Miscellaneous updates to comment text

* Rework the `get_depot_tools()` function into `get_google_repo()`, so that we can use it to pull down both the `depot_tools` and `build` repositories.

* More consistent handling of function arguments

* Reformatting of various commands for better legibility, as well as more generous use of whitespace.

* Add run hooks for LLVM and Rust elements. The bulk of this is not used in our tarball, but we do need various bits from it to be able to match what Google has.

* Also add run hook to copy in `clang-format`.

* New function `prune_lite_excluded_dirs()` that matches what the `export_lite_tarball()` function in `publish_tarball.py` does (and the reason why we have `check.patch`, since we need to keep this in sync)

* Invoke the upstream `export_tarball.py` script in situ instead of our own copy.

* Invoke the upstream `generate_hashes.py` script instead of our re-creation.

* Use `check.patch` to check for breaking changes in the upstream scripts.

* Temporarily patch scripts in the `build` repo and main source tree for our purposes (patches described below).

* Set an environment variable, implemented by one of the aforemention patches, to reduce the size of the LLVM and Rust clone operations.

* Prune the source tree for lite-ness prior to exporting tarballs.

New file `tweak-src.patch`: Patches `src/tools/clang/scripts/build.py` in the main source tree to implement `EXTRA_GIT_CLONE_ARGS` and skip the downloading of the Debian sysroots, all to reduce the size of the LLVM/Rust Git pulls.

New file `tweak-tools.patch`: Patch the upstream `export_tarball.py` script with a hack to make the `tarfile` module produce bit-identical output to Google's.

Kangie commented:
Ack, will review & test. Thanks for your work!

Kangie commented:
Overall this is fantastic; no nitpicks! Well documented and after examining the changes there's nothing that I had to really think about the 'why' of.

Testing (at least the comparing of hashes) will have to wait... I made the mistake of building a tarball that upstream haven't published yet! I'd test on a different tag but I'd prefer to have use of my internet connection for the next hour or so. :wink:

I do have one concern (that we can probably address in the longer term; it shouldn't hold up better tarballs now):

https://github.com/chromium-linux-tarballs/chromium-tarballs/blob/6b8e262cd07273a44f9e97d35aa5421a9f13302d/package_chromium.sh#L102-L110

The above is what broke upstream's CI process in the first place and necessitated the creation of separate tarball generation infrastructure. While some of the delay was down to upstream CI being full of jobs that could never run, constantly being resubmitted, there was a period where the Goo^H^H^HChromium patchset did not apply to Rust, bringing the whole process to a screeching halt at that point in the build.

Since _not_ packaging it would conflict with our new goal of producing bit-for-bit identical tarballs on standalone infrastructure, perhaps we can work out some way to have them packaged separately? Upstream already ship the output of their LLVM and Rust builds for `x86_64` and `aarch64` (on macOS only...), so it may make sense to provide the pre-processed sources for them so that those components can be reproduced if desired, eliminating the need to run that post-checkout hook for a lite tarball (or enabling the directory to be pruned).

I'm happy for you to go ahead and merge this when you're comfortable, maybe with a followup commit to update the README? I look forward to _even smaller_ tarballs!

I'll report back on if the hashes match when upstream tarballs for 134.0.6988.1 are out, but there's really no reason to suspect that they won't.

iskunk commented:
> Testing (at least the comparing of hashes) will have to wait... I made the mistake of building a tarball that upstream haven't published yet! I'd test on a different tag but I'd prefer to have use of my internet connection for the next hour or so. 😉

Be sure to give it a try, even if running overnight. It's a pretty neat trick! I've confirmed a match for the 132.0.6834.159 tarballs.

> The above is what broke upstream's CI process in the first place and necessitated the creation of separate tarball generation infrastructure. While some of the delay was down to upstream CI being full of jobs that could never run, constantly being resubmitted, there was a period where the Goo^H^H^HChromium patchset did not apply to Rust, bringing the whole process to a screeching halt at that point in the build.

As I understand, all those two commands are doing is a `git clone` operation of the LLVM and Rust trees, switching to a specific revision thereof, and then recursively shallow-cloning all the submodules. (There is a step where it downloads the Debian sysroots, but I patched that out as the sysroots don't make it into the tarball at all.) I'm not aware of the scripts patching any files... if you look at the Git trees after making a tarball, the only modified file you see is the README for GN.

> Since _not_ packaging it would conflict with our new goal of producing bit-for-bit identical tarballs on standalone infrastructure, perhaps we can work out some way to have them packaged separately? Upstream already ship the output of their LLVM and Rust builds for `x86_64` and `aarch64` (on macOS only...), so it may make sense to provide the pre-processed sources for them so that those components can be reproduced if desired, eliminating the need to run that post-checkout hook for a lite tarball (or enabling the directory to be pruned).

Note that the LLVM and Rust directories *are* pruned. What happens is that "pruning" is not a straight `rm -rf`; there are two exceptions to it:

1. Any GN-related files are kept, along with a shortlist of others (since deleting them has a tendency to break `gn gen` even if you don't actually need them for the build);
2. The (empty) directory structure is kept, because the `export_tarball` script can't recognize empties before `tarfile` recurses into them.

I'm not running the LLVM/Rust hooks to include those components in our tarball. I'm running the hooks to include *the post-pruning remains* of those components, so that we can match Google's lite tarball.

It's possible that the `llvm` and `rust-src` directories could be special-cased in the upstream `publish_tarball` script, so they are pruned with `rm -rf` instead of the big `find` expression. That would eliminate the need for us to run the associated hooks---and I considered doing that. But that would require a pretty hard guarantee that those subtrees won't ever have any GN/etc. files that are read from the top level of the Chromium tree. And between trying to ascertain that, versus just running the hooks, I figured running the hooks wasn't *that* big a deal.

> I'm happy for you to go ahead and merge this when you're comfortable, maybe with a followup commit to update the README? I look forward to _even smaller_ tarballs!

Yep, that sounds good, just let me know that the process works correctly for you.

(The README file also needs a link to the current archive location)

> I'll report back on if the hashes match when upstream tarballs for 134.0.6988.1 are out, but there's really no reason to suspect that they won't.

:+1: 

> EDIT: As a stretch goal, since the files are _well_ below the 2G threshold, we could upload the additional files as GH release artefacts (in addition to S3), so that everything consumers need is in one place. The S3 usage is quite reasonable atm (especially behind the CDN), but it's not sponsored or otherwise guaranteed. It's not essential and could easily be a follow-up Issue/PR

Yes, I think we are now in a good position to host the tarballs on GitHub as release artifacts.

We may want to reconsider the practice of keeping a back catalog, since if we are generating tarballs that exactly match Google's, then there isn't much of a point in providing what are effectively copies.

Kangie commented:
Looking good, hashes for 134.0.6990.0 match. I do note that the upstream -lite hashes have the wrong filename in them....

```
$ curl -L https://gsdview.appspot.com/chromium-browser-official/chromium-134.0.6990.0-lite.tar.xz.hashes
md5  c3b7270a810723e26a077a9560cef791  chromium-134.0.6990.0.tar.xz
sha1  9bbc336477b5511fe1dece25e7eb488ca305398e  chromium-134.0.6990.0.tar.xz
sha224  ed1bdba485483f62f7b6bb6f8fbf3076ca42976b2865aa4146110e41  chromium-134.0.6990.0.tar.xz
sha256  00d606bc07668d96118054adf575ebfda21b046da79f09c5b0c0428c3edf6bde  chromium-134.0.6990.0.tar.xz
sha384  9cc8bd73379eb486cc87e154e82d87f032d28bf9d8d5d2fbe70bcec4d24f5917d1f5c05ed1e2b2f554ff061ab0fda5c6  chromium-134.0.6990.0.tar.xz
sha512  a39628c2b5a9ebe8fe4746e52d48d7af6ed4e10e6b55035406046159f3c01110530668139939a24b9a7bf5381d5d4ae10c712391623938ef1aa4b907951335c2  chromium-134.0.6990.0.tar.xz
```

iskunk commented:
> This commit removes the cherry-pick that was causing issues; the Clang script has a function that can be used similarly, though isn't (currently) called.
> 
> [chromium/chromium@6bd5285](https://github.com/chromium/chromium/commit/6bd5285e4a8222d720a795180d817dc6fb7f3f06)
> 
> And removing them entirely until they need the next ones:
> 
> [chromium/chromium@4983450](https://github.com/chromium/chromium/commit/49834507987ef120b8733b14ec17f34d3118d244)
> 
> I'd like to isolate us (and ideally upstream, at least in terms of 'Chromium Sources') from this breakage, though if it ever does come up we can also just not run that hook at the cost of not being identical.
> 
> Upstream never published that hash, so I've kicked off an overnight run to compare hashes.

Okay, I see the [failed CI run](https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8734357732302229249/+/u/package_rust/stdout) where the conflict occurred. But did that issue ever make it into a tagged release? (If there is a version where this breakage occurs, I can try to put in a workaround in the script.) Otherwise, it seems like any other issue that their CI processes would catch prior to release.

> Looking good, hashes for 134.0.6990.0 match. I do note that the upstream -lite hashes have the wrong filename in them....

I don't understand why, but if you look at the bottom of the `export_lite_tarball()` function in `publish_tarball.py`, you'll see it's pretty explicit about that. I can try filing an upstream change for it...

Kangie commented:
> Okay, I see the [failed CI run](https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8734357732302229249/+/u/package_rust/stdout) where the conflict occurred. But did that issue ever make it into a tagged release?

Yes, this impacted upstream's Chromium tarball generation from 11 Oct 24 to Mid-November, impacting all channels through that period (as promotions happened, etc). See the following 'Failed CI runs' threads (in batches of 1000):
- [Thread 1](https://groups.google.com/a/chromium.org/g/chromium-packagers/c/9X0k0IXCkKY)
- [Thread 2](https://groups.google.com/a/chromium.org/g/chromium-packagers/c/J-5D8vrBOEU)

Here's a full list of impacted tags where release tarballs _cannot_ be generated due to the Rust hook:
https://chromium.googlesource.com/chromium/tools/build/+/refs/heads/main/recipes/recipes/publish_tarball.py#67

While not all of the ~2000 failed CI runs represent unique releases (due to retries), a significant number were affected.  This led to delays and required manual intervention, including blacklisting M131 and M132 tags upstream and the creation of this repository (see [Thread 3](https://groups.google.com/a/chromium.org/g/chromium-packagers/c/wjv9UKg2u4w) for my initial "I had to publish these tarballs").  Interestingly, some affected releases still made it to the Chrome Release Blog, suggesting workarounds were in place for the official Google Chrome builds.

The upstream "Solution" was to explicitly blacklist most of M131 tags after the 'beta' cut, and all of impacted M132 'dev' tags (as per `publish_tarball.py` above) so that they won't be picked up and run on CI.

> Otherwise, it seems like any other issue that their CI processes would catch prior to release.

I'm not so sure, given the above >100 tags that cannot be generated. If they were going to catch something like this I'd expect maybe 1 or 2 revisions, at worst. Ideally tagging a release would indicate that it's actually passed testing (etc), but tags get made arbitrarily upstream. There's no actual guarantee that a given tag actually passes CI (although they're _supposed_ to keep `main` in a buildable state to branch off). I assume that the official Release tag passes Google Chrome CI however as discussed below that doesn't really help us.

I don't know how they were still getting Google Chrome releases out in the meantime, though I'll outline my theory below:

Google don't notice this because the internal Chrome release process is separate from Chromium's, relying on builds from `git` entirely, likely using a completely separate set of scripts and CI infrastructure.

There's some evidence for this, in the form of a rejected ticket to build binary releases from release tarballs due to issues with `svn` checkouts. Google instead opted to switch to using `git` years ago (and closed the 'use tarballs for CI builds' crbug in response).

Google don't  seem to _really_ care about Chromium on Linux, AFAICT. Upstream doesn't produce Linux binaries like they do for MacOS and Windows, and it can be challenging to get eyes on a legitimate bug or regression in a timely manner (I have one outstanding for M133 that's been in for nearly a month, is now in the `stable` channel, and has been assigned against the appropriate component).

I don't think this should delay the merge of this PR - As you've mentioned we can work around it downstream if it recurs. Perhaps we can discuss generating the Rust and LLVM tarballs separately (at least for -lite) with upstream, and implement it in a manner that doesn't block the publication of Chromium tarballs so that at the very least the core Chromium code goes out in a '-lite' tarball for a given tag.

P.S.: If you're looking for more Chromium "Fun" I can point you at [this crbug](https://issues.chromium.org/issues/393183848), my attempt to make it possible to actually ship `beta` and `dev` channels without impacting on `stable`.

iskunk commented:
I dug into those links you gave, and experimented with the failing versions (132.0.6793.3 breaks as expected). I've updated the script (and PR) not to error out on failures in the LLVM and Rust steps---it will simply print a warning, delete the associated directory, and continue on. There is no distinction between a Git conflict issue versus some other failure mode (e.g. network flakiness) due to lack of a good way of telling them apart, but this should at least address the concern with those two steps stopping the show.

Let me know what you think. I also added a bit to the workflow to better indicate when this scenario occurs.

> Google don't seem to _really_ care about Chromium on Linux, AFAICT. Upstream doesn't produce Linux binaries like they do for MacOS and Windows

Don't they provide binary `.deb`/`.rpm` files of Chrome?

I think they care less about tarballs per se than about Linux generally, though the latter would still rank well below Windows and MacOS, given the size of the userbase. But I don't see this as a major issue, given that they've been responsive in merging my updates to their tarball-generation process, and that we can reproduce the same tarballs bit-perfectly.

There may be room for some changes that make it unnecessary for us to run the LLVM and Rust hooks at all, but I still need to do some investigation. I already have a fix for the issue of incorrect filenames in the `.hashes` files.

> P.S.: If you're looking for more Chromium "Fun" I can point you at [this crbug](https://issues.chromium.org/issues/393183848), my attempt to make it possible to actually ship `beta` and `dev` channels without impacting on `stable`.

Could you get the same effect with a wrapper script that passes an appropriate `--user-data-dir=...` option?

iskunk commented:
> I'm happy with the changes overall - We can tidy up Rust / LLVM later on if we get buy-in upstream.

All right, I've merged this in. Let me know if anything seems out of place!

> > Don't they provide binary .deb/.rpm files of Chrome?
> 
> Yes, I specifically meant Chromium, sorry!

Ah, I see what you mean. Official end-user binaries of Chromium (as opposed to Chrome) already exist in a penumbra between non-technical users and developers. There's not a whole lot of reason for them to exist in the first place, even for Win/Mac, since Google would rather everyone ran Chrome. No doubt the argument for putting them up was some variant of "well, we're already getting these binaries for free with our CI builds..."

Kangie commented:
> All right, I've merged this in. Let me know if anything seems out of place!

Will do. I look forward to even smaller downloads when `stable` gets bumped shortly!
